ql/src/java/org/apache/hadoop/hive/ql/plan/TezEdgeProperty.java

```diff
@@ -51,19 +53,22 @@ public TezEdgeProperty(HiveConf hiveConf, EdgeType edgeType,
   }
 
   public TezEdgeProperty(HiveConf hiveConf, EdgeType edgeType, boolean isAutoReduce,
-      boolean isSlowStart, int minReducer, int maxReducer, long bytesPerReducer) {
+      boolean isSlowStart, int minReducer, int maxReducer, long bytesPerReducer,
+      float minSrcFraction, float maxSrcFraction) {
     this(hiveConf, edgeType, -1);
-    setAutoReduce(hiveConf, isAutoReduce, minReducer, maxReducer, bytesPerReducer);
+    setAutoReduce(hiveConf, isAutoReduce, minReducer, maxReducer, bytesPerReducer, minSrcFraction, maxSrcFraction);
     this.isSlowStart = isSlowStart;
   }
 
   public void setAutoReduce(HiveConf hiveConf, boolean isAutoReduce, int minReducer,
-      int maxReducer, long bytesPerReducer) {
+      int maxReducer, long bytesPerReducer, float minSrcFraction, float maxSrcFraction) {
     this.hiveConf = hiveConf;
     this.minReducer = minReducer;
     this.maxReducer = maxReducer;
     this.isAutoReduce = isAutoReduce;
     this.inputSizePerReducer = bytesPerReducer;
+    this.minSrcFraction = minSrcFraction;
+    this.maxSrcFraction = maxSrcFraction;
```

Comment:
I think [the first commit](https://github.com/apache/hive/pull/4404/commits/c87b4ca6652dffc3cc150e54c793477108d8a74f) is the simplest.
However, Gopal suggested I should move to the current style while the previous PR is open.
https://issues.apache.org/jira/browse/HIVE-24485?focusedCommentId=17278522&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17278522

I personally don't have very strong opinions about how to support the params.

Reply:
@okumin : thanks for working on this, I believe the latest patch with Gopal's suggestion looks good, and I feel we can support all (currently 5) properties like this:
https://github.com/apache/tez/blob/ccad71dabceb850b58ff38e77fd206e68f62ce4b/tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/ShuffleVertexManager.java#L71-L110

Reply:
@abstractdog Thanks for putting your opinion. Actually, all the other 3 params are already covered indirectly. The newly added 2 are only ones we can't control at all.
https://github.com/apache/hive/blob/b55f9c513b25e8e61150f6bf0afd1a882780d098/ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java#L1657-L1669

- `tez.shuffle-vertex-manager.enable.auto-parallel`: This is configured when `hive.tez.auto.reducer.parallelism` is enabled and the reduce vertex can be parallelized
- `tez.shuffle-vertex-manager.desired-task-input-size`: This is inherited from `hive.exec.reducers.bytes.per.reducer`
- `tez.shuffle-vertex-manager.min-task-parallelism`: This will be `{estimated # of reducers} * hive.tez.min.partition.factor`

If we support them, it means `tez.shuffle-vertex-manager.*` overwrites the 3 params. Is it preferable?
Another option is to introduce new params like `hive.tez.auto.reducer.parallelism.min-src-fraction` assuming `tez.shuffle-vertex-manager.*` is SPI-like params. It might be consistent with other params.